v1/tarball/image.go

```diff
@@ -15,17 +15,116 @@
 package tarball
 
 import (
+	"archive/tar"
+	"bytes"
 	"fmt"
+	"io"
+	"os"
+
+	"github.com/google/go-containerregistry/v1/types"
 
 	"github.com/google/go-containerregistry/v1"
 )
 
-// TODO(mattmoor): This.
-// // image accesses an image from a remote registry
-// type image struct{}
-// var _ v1.Image = (*image)(nil)
+type image struct {
+	path     string
+	manifest *v1.Manifest
+}
+
+var _ v1.Image = (*image)(nil)
 
 // Image exposes an image from the tarball at the provided path.
 func Image(p string) (v1.Image, error) {
-	return nil, fmt.Errorf("NYI: tarball.Image(%v)", p)
+	img := image{
+		path: p,
+	}
+	return &img, nil
+}
+
+func (i *image) FSLayers() ([]v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) DiffIDs() ([]v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) ConfigName() (v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) BlobSet() (map[v1.Hash]struct{}, error) {
+	panic("not implemented")
+}
+
+func (i *image) Digest() v1.Hash {
+	panic("not implemented")
+}
+
+func (i *image) MediaType() types.MediaType {
+	panic("not implemented")
+}
+
+// Manifest returns the v1.Manifest for the specified image.
+// This method memoizes the result, avoiding repeated reads from the tarball.
+func (i *image) Manifest() (*v1.Manifest, error) {
+	if i.manifest == nil {
+		f, err := os.Open(i.path)
+		defer f.Close()
+		if err != nil {
+			return nil, err
+		}
+		m, err := loadManifest(f)
+		if err != nil {
+			return nil, err
+		}
+		i.manifest = m
+	}
+	return i.manifest, nil
+}
+
+func loadManifest(r io.Reader) (*v1.Manifest, error) {
+	tf := tar.NewReader(r)
+	for {
+		hdr, err := tf.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		var buf bytes.Buffer
+		if hdr.Name == "manifest.json" {
+			_, err := io.Copy(&buf, tf)
+			if err != nil {
+				return nil, err
+			}
+			return v1.ParseManifest(buf.Bytes())
```

Comment:
Sadly, this is not the manifest you are looking for :(

Reply:
Ah whoops, I was hoping I could reuse the type for `Config` and `Layers`, but it looks like those `Descriptors` are too strongly-typed for that :)